src/util/heap/freelistpageresource.rs

```diff
@@ -210,42 +167,25 @@ impl<VM: VMBinding> FreeListPageResource<VM> {
     }
 
     pub fn new_discontiguous(vm_map: &'static dyn VMMap) -> Self {
-        let common_flpr = {
-            let start = vm_layout().available_start();
-            let common_flpr = Box::new(CommonFreeListPageResource {
-                free_list: vm_map.create_freelist(start),
-                start,
-            });
-            // `CommonFreeListPageResource` lives as a member in space instances.
-            // Since `Space` instances are always stored as global variables, so it is okay here
-            // to turn `&CommonFreeListPageResource` into `&'static CommonFreeListPageResource`
-            unsafe {
-                vm_map.bind_freelist(common_flpr.as_ref() as *const _);
-            }
-            common_flpr
-        };
+        let start = vm_layout().available_start();
+        let free_list = vm_map.create_freelist(start);
+        debug_assert!(
+            free_list.downcast_ref::<RawMemoryFreeList>().is_none(),
```

Comment:
Same here. We could let `dyn FreeList` to tell us if discontiguous space is allowed.

Reply:
Like I said, I don't think `FreeList` should be aware of whether it is used by one kind of space or another.

Our current implementation happens to disallow discontiguous spaces when using `Map64`, and `Map64` happens to select `RawMemoryFreeList`.  I think we can remove this assertion because it's only the status quo, and there is no reason why we can't use `RawMemoryFreeList` in discontiguous spaces.

Reply:
After a second thought, I think we should leave the comment as is, because it is an implementation detail of the current `FreeListPageResource`.  The way we deal with discontiguous space is that the `start` variable here is meaningless.  But when the address range of the discontiguous space is determined, we re-assign the `address` of all discontiguous page resources.  (That's the callback you saw in `mmtk.rs`.)  If `start` needs an offset to make room for the `RawMemoryFreeList`, it will be overwritten.

We could adapt to that possibility by requiring `start` to be initially zero (when using `IntArrayFreeList`) or the offset (when using `RawMemoryFreeList`) for discontiguous spaces.  When the discontiguous range is determined, we *add* the starting address to the discontiguous range to the `start` variable instead of overwriting it.  But we have no way to test that because we never have that use case, and I already proposed to revamp the way we create spaces (see [here](https://github.com/mmtk/mmtk-core/issues/932) and [here](https://github.com/mmtk/mmtk-core/issues/984)).  So we can leave this assertion here to reflect the status quo.  `downcast_ref` is not as elegant as virtual methods from OOP's point of view, but we are just inspecting an implementation detail, and it is in an assertion.  I'll add comment to clarify that this is only the status quo, not something that has to be true in general.

Reply:
I added comments to this assertion.